mypy/checkexpr.py

```diff
@@ -1609,10 +1608,17 @@ def analyze_super(self, e: SuperExpr, is_lvalue: bool) -> Type:
                         return AnyType()
                     if not self.chk.in_checked_function():
                         return AnyType()
-                    return analyze_member_access(e.name, self_type(e.info), e,
-                                                 is_lvalue, True, False,
-                                                 self.named_type, self.not_ready_callback,
-                                                 self.msg, base, chk=self.chk)
+                    # fill_typevars(e.info) erases type variables
+                    erased_self = fill_typevars(e.info)
+                    args = self.chk.function_stack[-1].arguments
+                    # An empty args with super() is an error, but we need something in declared_self
+                    declared_self = args[0].variable.type if args else erased_self
+                    return analyze_member_access(name=e.name, typ=erased_self, node=e,
```

Comment:
Yeah, it could well break for generic classes, but as we don't have tests for them we don't know whether they work right now. Let's do that separately. Even for generic classes this is perhaps not the best approach.